smart.8
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" Copyright (c) 2021-2026 Chuck Tuffli
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" Note: The date here should be updated whenever a non-trivial
.\" change is made to the manual page.
.Dd February 14, 2026
.Dt SMART 8
.Os
.Sh NAME
.Nm smart ,
.Nm diskhealth
.Nd monitor disk health from a storage device via SMART
.Sh SYNOPSIS
.Nm
.Op Fl dhitvx
.Oo Fl a Ar page:attribute Ns Oo , Ns Ar page:attribute Oc Ns ... Oc
.Op Fl Fl debug
.Ar device
.Nm diskhealth
.Op Fl Dhitvx
.Oo Fl a Ar page:attribute Ns Oo , Ns Ar page:attribute Oc Ns ... Oc
.Op Fl Fl debug
.Ar device
.Sh DESCRIPTION
The
.Nm
command allows the user to monitor the various information reported
by Self-Monitoring, Analysis and Reporting Technology (SMART) present
on most ATA, SCSI, and NVMe storage media.
Because the structure of this information varies across protocols,
.Nm
normalizes entries using the format:
.Bd -literal
	<Page ID>    <Attribute ID>    <Value>    <Thresholds>
.Ed
.Pp
Fields are tab-delimited by default, but the command can output
data in any format supported by
.Xr libxo 3 .
.Pp
Because ATA does not have log pages,
.Nm
uses the Command Feature field value in place of the log page ID.
For SMART READ DATA, this value is 208 / 0xd0.
Note that devices choose which attribute ID values they support, their
descriptions, and the format of the data.
The three values displayed with the
.Fl Fl threshold
option are:
.Pp
.Bl -dash -compact -offset indent
.It
Status flags (byte offset 1, 2 bytes)
.It
Nominal attribute value (byte offset 3, 1 byte)
.It
Worst Ever attribute value (byte offset 4, 1 byte)
.El
.Pp
Additionally,
.Nm
reports the value of the SMART STATUS command (Command Feature field
value 218 / 0xda).
As this command does not return any data,
the command represents this entry with a synthetic attribute
ID of 0, and it uses the command status (0 or 1) as the attribute
value.
.Pp
NVMe devices support the SMART/Health log page (Page ID 2 / 0x2).
The data returned in this log page is not structured as attribute IDs.
Instead,
.Nm
uses the byte offset of each field as the attribute ID.
For example,
byte 3 is the Available Spare.
Thus, for NVMe, attribute ID 3 is
Available Spare.
Note that NVMe health data does not include threshold
values, and as a result, the command will ignore the
.Fl Fl threshold
option.
.Pp
SCSI devices can support a number of log pages which report drive
health.
The command will report the following pages:
.Pp
.Bl -dash -compact -offset indent
.It
Write Errors (Page ID 2 / 0x2)
.It
Read Errors (Page ID 3 / 0x3)
.It
Verify Errors (Page ID 5 / 0x5)
.It
Non-medium Errors (Page ID 6 / 0x6)
.It
Last N Errors (Page ID 7 / 0x7)
.It
Temperature (Page ID 13 / 0xd)
.It
Start-stop Cycles (Page ID 14 / 0x3)
.It
Informational Exceptions (Page ID 47 / 0x2f)
.El
.Pp
Note that all log pages are optional, and a particular drive
may not support all these pages.
For SCSI devices, the Attribute ID
maps to the SCSI parameter code defined by the command.
Parameter
codes are integer values from 0 to N, and, by themselves, are ambiguous
outside the context of a particular log page.
Note that SCSI health data
does not include threshold values, and as a result, the command will
ignore the
.Fl Fl threshold
option.
.Pp
The following options are available:
.Bl -tag -width "-d argument"
.It Fl a Ar page:attribute , Fl Fl attribute= Ns Ar page:attribute
A comma-separated list of attributes to display.
If page is missing, display the matching attribute from any page.
.It Fl d , Fl Fl decode
Decode the attribute ID values.
This is the default when invoked as
.Nm diskhealth .
.It Fl D , Fl Fl no-decode
Do not decode the attribute ID values.
This is the default when invoked as
.Nm .
.It Fl h , Fl Fl help
Prints a usage message and exits.
.It Fl i , Fl Fl info
Print general device information.
.It Fl t , Fl Fl threshold
Also print the threshold values.
.It Fl v , Fl Fl version
Print the version and copyright.
.It Fl x , Fl Fl hex
Print the values in hexadecimal.
.It Ar device
An explicit device path
.Pq Pa /dev/ada0
or GEOM provider
.Pq Pa ada0
.
.El
.El
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
Print all SMART READ DATA and SMART STATUS including the
threshold values for ATA drive ada0.
.Pp
.Dl # smart -t ada0
.Pps
.Pp
Print only attribute ID 5 ("Reallocated Sectors Count") for
ATA drive ada0.
.Pp
.Dl # smart -a 5 ada0
.Pps
.Pp
Print attribute IDs 5 ("Reallocated Sectors Count") and 171
("SSD Program Fail") for ATA drive ada0.
.Pp
.Dl # smart -a 5,171 ada0
.Pps
.Pp
Print all health pages supported by SCSI device da0 including:
.Bl -dash -compact -offset indent
.It
Write Errors
.It
Read Errors
.It
Verify Errors
.It
Non-medium Errors
.It
Last N Errors
.It
Temperature
.It
Start-stop Cycles
.It
Informational Exceptions
.El
.Pp
.Dl # smart da0
.Pps
.Pp
Print all Health log page entries in hexadecimal for NVMe
device nda0.
.Pp
.Dl # smart -x nda0
.Pps
.Sh DIAGNOSTICS
The command may fail for one of the following reasons:
.Bl -diag
.It "No output displayed"
The device does not support health data.
.It "CAMGETPASSTHRU ioctl failed"
.Nm
relies on
.Xr cam 4
to retrieve data from devices and will display this message if the
device does not have a passthrough driver.
This can happen, for
example, if the system uses the
.Xr nvd 4
NVMe driver instead of the
.Xr nda 4
driver.
.El
.Sh SEE ALSO
.Xr libxo 3 ,
.Xr cam 4 ,
.Xr nda 4 ,
.Xr nvd 4
.Sh AUTHORS
This
utility was written by
.An Chuck Tuffli Aq Mt chuck@FreeBSD.org .
.Sh BUGS
Probably.


Changelog
This file documents changes for smart releases

version 1.0.0
 - Fix ATA threshold output (gh-10). This is a breaking change as it
   reduces the output from 4 fields to 3 (drops the "reserved" byte
   from threshold).
 - Fix the ATA raw output. This is a breaking change as it increase the
   output from 6 bytes to 7 (i.e., includes the "reserved" byte). Note
   that while some attributes use this byte, most do not.
 - Fix direct debug output (--debug) to standard error
 - Use POSIX memcpy and memset instead of older bXXX equivalents

version 0.4.2
 - Update README contents

version 0.4.1
  - Allow a comma-separated list of attributes
  - Code refactor + update code comments

version 0.3.0

  - Reclaim the -d option from debug
  - Change field separator from spaces to tab
  - Add textual descriptions of attribute IDs for ATA, NVMe, and SCSI
  - Add a manual page
  - Fixes
     * libxo structure for attribute and attributes
     * simplify LIBXO ifdef sprawl
     * display of threshold values
     * display of long values


smart.c
/*
 * Copyright (c) 2016-2026 Chuck Tuffli <chuck@tuffli.net>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <stdbool.h>
#include <string.h>
#ifdef LIBXO
#include <libxo/xo.h>
#endif

#include "libsmart.h"

#define SMART_NAME "smart"
#define SMART_VERSION	"1.0.0"

extern bool do_debug;

static const char *pn;
bool do_debug = false;
static int debugset = 0;

static struct option opts[] = {
	{ "help", no_argument, NULL, 'h' },
	{ "threshold", no_argument, NULL, 't' },
	{ "hex", no_argument, NULL, 'x' },
	{ "attribute", required_argument, NULL, 'a' },
	{ "info", no_argument, NULL, 'i' },
	{ "version", no_argument, NULL, 'v' },
	{ "decode", no_argument, NULL, 'd' },
	{ "no-decode", no_argument, NULL, 'D' },
	{ "debug", no_argument, &debugset, 1 },
	{ NULL, 0, NULL, 0 }
};

static void
usage(const char *name)
{
	printf("Usage: %s [-htxi] [-a attribute[,attribute]...] <device name>\n", name);
	printf("\t-h, --help\n");
	printf("\t-t, --threshold : also print out the threshold values\n");
	printf("\t-x, --hex : print the values out in hexadecimal\n");
	printf("\t-a, --attribute : print specified attribute(s)\n");
	printf("\t-i, --info : print general device information\n");
	printf("\t-d, --decode: decode the attribute IDs\n");
	printf("\t-D, --no-decode: don't decode the attribute IDs\n");
	printf("\t-v, --version : print the version and copyright\n");
	printf("\t    --debug : output diagnostic information\n");
}

/*
 * Convert string to an integer
 *
 * Returns -1 on error, converted value otherwise
 */
static int32_t
get_val(char *attr, char **next)
{
	char *sep = NULL;
	long val;

	*next = NULL;

	val = strtol(attr, &sep, 0);
	if ((val == 0) && (errno != 0)) {
		printf("Error parsing attribute %s", attr);
		switch (errno) {
		case EINVAL:
			printf(" (not a number?)\n");
			break;
		case ERANGE:
			printf(" (value out of range)\n");
			break;
		default:
			printf("\n");
		}
		return -1;
	}

	if (val > INT32_MAX) {
		printf("Attribute value %ld too big\n", val);
		return -1;
	}

	*next = sep;
	return ((int32_t)val);
}

/*
 * Create a match specification from the given attribute
 *
 * Attribute format is
 *     <Page ID>:<Attribute ID>
 * where page and attribute IDs are integers. If the page ID is missing,
 * match the specified attribute ID on any page (i.e. -1). Valid forms are
 *    <int>:<int>
 *    :<int>
 *    <int>
 *
 * Returns 0 on success
 */
static int
add_match(smart_matches_t **matches, char *attr)
{
	char *next;
	int32_t page = -1, id;
	uint32_t count = 0;

	id = get_val(attr, &next);
	if (id < 0)
		return id;

	if (*next == ':') {
		page = id;
		id = get_val(next + 1, &next);
		if (id < 0)
			return id;
	}

	if (*matches == NULL) {
		*matches = calloc(1, sizeof(smart_matches_t) + sizeof(smart_match_t));
		if (*matches == NULL)
			return ENOMEM;
	} else {
		void *tmp;

		count = (*matches)->count;
		tmp = realloc(*matches, sizeof(smart_matches_t) + ((count + 1) * sizeof(smart_match_t)));
		if (tmp == NULL)
			return ENOMEM;
		*matches = tmp;
	}

	(*matches)->m[count].page = page;
	(*matches)->m[count].id = id;
	(*matches)->count++;
	return 0;
}

/*
 * Parse the comma separated list of attributes to match
 *
 * Caller frees memory allocated for the smart_matches_t pointer.
 *
 * Returns 0 on success
 */
static int
parse_matches(smart_matches_t **matches, char *attr)
{
	int res;

	if (attr[0] == '\0')
		return -1;

	while (*attr != '\0') {
		char *next;
		size_t len;

		if ((next = strchr(attr, ',')) == NULL) {
			len = strlen(attr);
			next = attr + len;
		} else {
			len = next - attr;
			next++;
		}

		if (len == 0) {
			printf("Malformed attribute %s\n", attr);
			return -1;
		}

		res = add_match(matches, attr);
		if (res)
			return res;

		attr = next;
	}

	return 0;
}

int
main(int argc, char *argv[])
{
	smart_h h;
	smart_map_t *sm = NULL;
	char *devname = NULL;
	int ch;
	bool do_thresh = false, do_hex = false, do_info = false, do_version = false,
	     do_descr;
	smart_matches_t *matches = NULL;
	int rc = EXIT_SUCCESS;

	/*
	 * By default, keep the original behavior (output numbers only) if
	 * invoked as smart. Otherwise, default to printing the human-friendly
	 * text descriptions.
	 */
	pn = getprogname();
	if (strcmp(pn, SMART_NAME) == 0)
		do_descr = false;
	else
		do_descr = true;

#ifdef LIBXO
	argc = xo_parse_args(argc, argv);
#endif

	while ((ch = getopt_long(argc, argv, "htxa:idDv", opts, NULL)) != -1) {
		switch (ch) {
		case 'h':
			usage(pn);
#ifdef LIBXO
			xo_finish();
#endif
			return EXIT_SUCCESS;
			break;
		case 't':
			do_thresh = true;
			break;
		case 'x':
			do_hex = true;
			break;
		case 'a':
			if (parse_matches(&matches, optarg)) {
				usage(pn);
				return EXIT_FAILURE;
			}
			break;
		case 'i':
			do_info = true;
			break;
		case 'd':
			do_descr = true;
			break;
		case 'D':
			do_descr = false;
			break;
		case 'v':
			do_version = true;
			break;
		case 0:
			if (debugset)
				do_debug = true;
			break;
		default:
			usage(pn);
#ifdef LIBXO
			xo_finish();
#endif
			return EXIT_FAILURE;
		}
	}

	if (do_version) {
		printf("%s, version %s\n", pn, SMART_VERSION);
		printf("Copyright (c) 2016-2026 Chuck Tuffli\n"
				"This is free software; see the source for copying conditions.\n");
		return EXIT_SUCCESS;
	}

	argc -= optind;
	argv += optind;

	devname = argv[0];

	if (!devname) {
		printf("no device specified\n");
		usage(pn);
#ifdef LIBXO
		xo_finish();
#endif
		return EXIT_FAILURE;
	}

	h = smart_open(SMART_PROTO_AUTO, argv[0]);

	if (h == NULL) {
		printf("device open failed %s\n", argv[0]);
#ifdef LIBXO
		xo_finish();
#endif
		return EXIT_FAILURE;
	}

#ifdef LIBXO
	xo_open_container("drive");
#endif

	if (do_info) {
		smart_print_device_info(h);
	}

	if (smart_supported(h)) {
		sm = smart_read(h);

		if (sm) {
			uint32_t flags = 0;

			if (do_hex)
				flags |= SMART_OPEN_F_HEX;
			if (do_thresh)
				flags |= SMART_OPEN_F_THRESH;
			if (do_descr)
				flags |= SMART_OPEN_F_DESCR;

			smart_print(h, sm, matches, flags);

			smart_free(sm);
		}
	} else {
		rc = EXIT_FAILURE;
	}
#ifdef LIBXO
	xo_finish();
#endif
	smart_close(h);

	return rc;
}
